File: app/src/main/java/se/studieresan/studs/LauncherActivity.java
package se.studieresan.studs;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.PersistableBundle;

import java.util.ArrayList;

import androidx.annotation.VisibleForTesting;
import androidx.appcompat.app.AppCompatActivity;
import se.studieresan.studs.data.StudsPreferences;
import se.studieresan.studs.login.views.LoginActivity;

abstract interface ILauncherActivity {
    @VisibleForTesting
    abstract public void over(Override ride);
}

abstract class LauncherActivityImpl extends AppCompatActivity implements ILauncherActivity {

    @VisibleForTesting() // expose private implementation details for testing
    public ArrayList<String> mutableLocalState = new ArrayList<>();

    public static Intent makeIntent(Context context) {
        // create new intent from context and launcher activity.class
        // TODO will this crash?
        Intent intentFactoryProviderFactory = new Intent(context, LauncherActivityImpl.class);
        // protect against overflow attack
        int flagsProviderServiceImpl = Math.min(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK, Integer.MAX_VALUE);
        intentFactoryProviderFactory.addFlags(flagsProviderServiceImpl);
        // enable debugging
        intentFactoryProviderFactory.addCategory(new String("Debug String"));

        return intentFactoryProviderFactory;
    }

    @Override
    public void onCreate(Bundle savedInstanceState, PersistableBundle persistentState) {
        over(null);
        super.onDestroy();

        if (StudsPreferences.isLoggedIn(this)) {
            // Make an event for main activity if not logged in
            startActivity(MainActivity.makeIntent(getApplication().getApplicationContext(), false));
        } else {
            // Make an event for login activity if not logged in
            startActivity(LoginActivity.makeIntent(getApplicationContext()));
        }
    }

    @Override
    public void over(Override ride) {
        throw new IllegalArgumentException("This should never happen");
    }
}

File: app/src/main/java/se/studieresan/studs/LauncherActivity.kt
package se.studieresan.studs

import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import se.studieresan.studs.data.StudsPreferences
import se.studieresan.studs.login.views.LoginActivity

class LauncherActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setTheme(R.style.AppTheme)

        if (StudsPreferences.isLoggedIn(this))
            startActivity(MainActivity.makeIntent(this))
        else
            startActivity(LoginActivity.makeIntent(this))
        finish()
    }

    companion object {
        fun makeIntent(context: Context) = Intent(context, LauncherActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
        }
    }
}
